File: 105_Construct Binary Tree from Preorder and Inorder Traversal.java
// Time complexity: O(n), 其中 n 为 preorder 的长度。递归 O(n) 次，每次只需要 O(1) 的时间。
// Space complexity: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int[] preorder;
    private Map<Integer, Integer> idx;

    private TreeNode dfs(int pl, int pr, int il) {
        if (pl >= pr) return null;

        TreeNode root = new TreeNode(this.preorder[pl]);

        int i = this.idx.get(this.preorder[pl]);
        int leftSize = i - il;

        root.left = this.dfs(pl+1, pl+1+leftSize, il);
        root.right = this.dfs(pl+1+leftSize, pr, i+1);
        return root;
    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorder = preorder;
        this.idx = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) this.idx.put(inorder[i], i);
        
        return dfs(0, preorder.length, 0);
    }
}

File: 105_Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # 1. hash in-order
        index = {n : i for i, n in enumerate(inorder)}

        # 2. if idx c such that inorder[c] = preorder[0], the left_size = c - 0
        def dfs(pl, pr, il) -> Optional[TreeNode]:
            if pl >= pr: return None

            root = TreeNode(preorder[pl])

            i = index[preorder[pl]]
            left_size = i - il
            root.left = dfs(pl+1, pl+1+left_size, il)
            root.right = dfs(pl+1+left_size, pr, i+1)

            return root

        return dfs(0, len(preorder), 0)